### PR TITLE
add cluster-scoped resources in NSTemplateTiers

### PR DESCRIPTION
**Pull request by xcoulon**
refactored logic to generate NSTemplateTiers, reading content
from generated Assets earlier in the process.
Also, refactored/reorganized tests related to NSTemplateTier generator
(moving/regrouping tests)
    
Also, moving namespace templates in a subdir to make room for
cluster resource quota templates, one per tier ('advanced'
and 'basic')
    
Fixes https://issues.redhat.com/browse/CRT-459
    
Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by xcoulon** [comment]
This PR is in `draft` mode until https://github.com/codeready-toolchain/api/pull/116 and https://github.com/codeready-toolchain/host-operator/pull/156 are merged, but it's ready for a first pass review and comments ;)

**Comment by xcoulon** [comment]
@alexeykazakov @MatousJobanek this PR is ready for review. 

Note: I will associate it with e2e tests to verify that the NSTemplateTier resources created at startup contain the new `spec.clusterResources` field

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go [hunk 1]
```diff
@@ -24,115 +24,189 @@ var log = logf.Log.WithName("templates")
...
+func loadClusterResourceQuotaAssets(get assetFunc) (map[string]asset, error) {
+	// get the revisions
+	metadata, err := get("metadata.yaml")
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load cluster resource quota templates")
+	}
+	revisions := make(map[string]string)
+	err = yaml.Unmarshal([]byte(metadata), &revisions)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to load cluster resource quota templates")
```
Comment (alexeykazakov) [inline pkg/templates/nstemplatetiers/nstemplatetier_generator.go]:
```suggestion
		return nil, errors.Wrapf(err, "unable to load cluster resource templates")
```

**Comment by alexeykazakov** [comment]
@xcoulon just FYI. These cluster resource quotas won't work until we update the namespace templates too. We need to set the default limits and requests there (see https://issues.redhat.com/browse/CRT-460). Otherwise the cluster quotas are not taken into account.
It's OK to do it in a separate PR though (as it defined in a separate JIRA). Just wanted to point out that it's critical to set the limits for the namespaces too to enforce the limits :)

**Comment by xcoulon** [comment]
> @xcoulon just FYI. These cluster resource quotas won't work until we update the namespace templates too. We need to set the default limits and requests there (see https://issues.redhat.com/browse/CRT-460). Otherwise the cluster quotas are not taken into account.
> It's OK to do it in a separate PR though (as it defined in a separate JIRA). Just wanted to point out that it's critical to set the limits for the namespaces too to enforce the limits :)

ok, thanks for the clarification, @alexeykazakov

File: deploy/templates/nstemplatetiers/clusterresources/advanced.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
+apiVersion: template.openshift.io/v1
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: yes, we could try something like that, indeed. Let me think about it

File: deploy/templates/nstemplatetiers/namespaces/metadata.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+advanced-code: "fd43e6b"
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/namespaces/metadata.yaml]: I probably committed this file before editing the `.gitignore` 🤦‍♂

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go [hunk 4]
```diff
@@ -24,115 +24,189 @@ var log = logf.Log.WithName("templates")
 
 // CreateOrUpdateResources generates the NSTemplateTier resources from the namespace templates,
 // then uses the manager's client to create or update the resources on the cluster.
-func CreateOrUpdateResources(s *runtime.Scheme, client client.Client, namespace string, asset func(name string) ([]byte, error)) error {
-	g, err := newNSTemplateTierGenerator(s, asset)
+func CreateOrUpdateResources(s *runtime.Scheme, client client.Client, namespace string, namespaceAssets assetFunc, clusterResourceQuotaAssets assetFunc) error {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: ok, if you like ;)

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go [hunk 5]
```diff
@@ -24,115 +24,189 @@ var log = logf.Log.WithName("templates")
...
 			return nil, errors.Errorf("invalid namespace template filename. Expected format: '<tier_kind>-<namespace_kind>', got %s", filename)
 		}
-		tierKind := data[0]
-		nsKind := data[1]
-		if _, ok := revisions[tierKind]; !ok {
-			revisions[tierKind] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage'
+		tier := data[0] // 'advanced', 'basic', etc.
+		typ := data[1]  // 'code', 'stage', etc.
+		if _, ok := revisionsByTier[tier]; !ok {
+			revisionsByTier[tier] = make(map[string]string, 3) // expect 3 entries: 'code', 'dev' and 'stage', but map size will increase if needed
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: that's the initial size of the map. It doesn't mean we will have exactly 3 entries.

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go [hunk 6]
```diff
@@ -24,115 +24,189 @@ var log = logf.Log.WithName("templates")
...
+
+// loadClusterResourceQuotaAssets returns a map in which:
+// - each key is a tier type (eg: "basic", "advanced", etc.)
+// - each value is an "asset", i.e., a template content ([]byte) and its revision (string)
+// Eg:
+// asset {
+//		revision: "abcdef",
+//		content: []byte{...},
+//	}
+func loadClusterResourceQuotaAssets(get assetFunc) (map[string]asset, error) {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: ah right, I forgot to rename this func 🤦‍♂ Thanks for catching that!

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: Sure, but the wording is confusing. We should not expect 3 entries. It can be any number.

[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: ok, so let me rephrase it, then

[the remark this answers is not in the document]
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: done in https://github.com/codeready-toolchain/host-operator/pull/157/commits/e7e771858b6a2a87de35ab2b4be018d382f00f5f

File: deploy/templates/nstemplatetiers/namespaces/metadata.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/namespaces/metadata.yaml]: done in https://github.com/codeready-toolchain/host-operator/pull/157/commits/e7e771858b6a2a87de35ab2b4be018d382f00f5f

File: deploy/templates/nstemplatetiers/clusterresources/advanced.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: or maybe we could do something like this:
```
/templates/tiers/advanced
    /namespaces
        /code.yaml
        /dev.yaml
        /stage.yaml
    /cluster
        /cluster.yaml
/templates/tiers/basic
    /namespaces
        /code.yaml
        /dev.yaml
        /stage.yaml
    /cluster
        /cluster.yaml
/templates/tiers/team
    /namespaces
        /code.yaml
        /dev.yaml
    /cluster
        /cluster.yaml
...
```
i.e., we keep the intermediate `namespaces` and `cluster` folders, so we know how to process the files and bind them to the `spec.namespaces` vs `spec.clusterResources` fields in the NSTemplateTier CRs?

[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: @alexeykazakov otherwise, we could keep your suggestion but rely on a (strict) file naming scheme to bind the templates to the spec.namespaces vs spec.clusterResources fields in the NSTemplateTier CRs.

**Comment by xcoulon** [comment]
Added e2e tests PR: https://github.com/codeready-toolchain/toolchain-e2e/pull/78

File: deploy/templates/nstemplatetiers/clusterresources/advanced.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: I would rather stay with what we have now or choose the Alexey's proposal. The additional directories for namespace and cluster templates under separated dirs for every tier looks really confusing and chaotic for me.

[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: > The additional directories for namespace and cluster templates under separated dirs for every tier looks really confusing and chaotic for me.
"chaotic", really? 😳
The problem with having all files in the same directory is how to distinguish the namespace-related templates vs the other cluster-scoped templates. For that, we could use a name pattern (eg: all namespace template filenames would have to start with `ns-`). Would it make it less confusing to you @MatousJobanek ?

[the remark this answers is not in the document]
Comment (alexeykazakov) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: I just thought that having all the (namespace) templates in one big dir is a bit messy. Especially when we have many tiers.
Ideally one dir per tier (and all the tier templates in that dir) would look better IMO.
And yes, for that we would need to use strict file name patterns to distinguish between cluster and ns templates. Which is fine.
Separate folders for cluster and ns would work too for me.
But let’s not spend too much time on it! I’m ok to keep it as is for now and refactor it later if it takes more than couple of hours for you to re-structure the dirs now :)

File: go.mod [hunk 7]
```diff
@@ -17,17 +22,25 @@ require (
...
 	k8s.io/client-go v12.0.0+incompatible
 	k8s.io/klog v1.0.0
+	k8s.io/kube-openapi v0.0.0-20200204173128-addea2498afe // indirect
+	k8s.io/utils v0.0.0-20200229041039-0a110f9eb7ab // indirect
 	sigs.k8s.io/controller-runtime v0.5.0
 	sigs.k8s.io/kubefed v0.1.0-rc6.0.20200224204536-6207193c49f7
+	sigs.k8s.io/yaml v1.2.0 // indirect
 )
 
+replace github.com/codeready-toolchain/api => github.com/alexeykazakov/api v0.0.0-20200306233032-6a4378dbcd57
```
Comment (alexeykazakov) [inline go.mod]: Will need to remove it when https://github.com/codeready-toolchain/api/pull/118 is merged.

Comment (xcoulon) [reply go.mod]: > @xcoulon the test were failing because the team tier introduced by Matous didn't have cluster resources so the generate failed to create that tier. I made a few changes:
> 
> * made cluster resource truly optional
> * added some tests for that ^
> * added cluster resources to the team tier. I think we should support tiers without cluster resources but in this particular case we do need to specify cluster resources for that tier.
> 
> Please review my changes!

I don't understand why we can't simply change the tests?

Also, maybe we could change the `ClusterResources NSTemplateTierClusterResources` field in the `NSTemplateTierSpec` type to `ClusterResources []NSTemplateTierClusterResources` (ie, an array) which means:
- we wouldn't have to use a pointer (I'm a bit reluctant to that, but that's my personal opinion) 
- we could have something more structured with subfolders as we discussed in https://github.com/codeready-toolchain/host-operator/pull/157/files#r388772370, with the possibility of multiple files in the `cluster` subdir in each "tier" folder.

WDYT @alexeykazakov and @MatousJobanek ?

File: deploy/templates/nstemplatetiers/clusterresources/advanced.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply deploy/templates/nstemplatetiers/clusterresources/advanced.yaml]: > For that, we could use a name pattern (eg: all namespace template filenames would have to start with ns-). Would it make it less confusing to you

yes, the name pattert would be fine, but I agree with Alexey:

> But let’s not spend too much time on it! I’m ok to keep it as is for now and refactor it later if it takes more than couple of hours for you to re-structure the dirs now :)

so keep it as it is and if we onboard more tiers, then we can think about seperating them :-)

**Comment by alexeykazakov** [comment]
> I don't understand why we can't simply change the tests?

Because it's not a test problem. It's a runtime problem. Our tests caught a bug.

**Comment by alexeykazakov** [comment]
> Also, maybe we could change the ClusterResources NSTemplateTierClusterResources field in the NSTemplateTierSpec type to ClusterResources []NSTemplateTierClusterResources (ie, an array) which means:

Technically it would work but IMO it would complicate our tests a bit. We have to support multiple NS templates because we have to support multiple namespaces. We don't really need to support multiple cluster templates though. But if we do then we have to cover it by tests. So we would need to tests:
- no cluster templates
- single cluster template
- multiple cluster templates

**Comment by xcoulon** [comment]
> > Also, maybe we could change the ClusterResources NSTemplateTierClusterResources field in the NSTemplateTierSpec type to ClusterResources []NSTemplateTierClusterResources (ie, an array) which means:
> 
> Technically it would work but IMO it would complicate our tests a bit. We have to support multiple NS templates because we have to support multiple namespaces. We don't really need to support multiple cluster templates though. But if we do then we have to cover it by tests. So we would need to tests:
> 
> * no cluster templates
> * single cluster template
> * multiple cluster templates

ok, in that case, I will go for your previous suggestion in https://github.com/codeready-toolchain/host-operator/pull/157#discussion_r388618596:
```
/templates/tiers/advanced/
  ns_dev.yaml
  ns_code.yaml
  ns_stage.yaml
  cluster.yaml
/templates/tiers/basic/
  ns_dev.yaml
  ns_code.yaml
  ns_stage.yaml
  cluster.yaml
/templates/tiers/team/
  ns_dev.yaml
  ns_stage.yaml
  cluster.yaml
...
```

**Comment by xcoulon** [comment]
replaced by https://github.com/codeready-toolchain/host-operator/pull/161
